database/RocksIterator.java

```diff
@@ -215,8 +215,8 @@ synchronized boolean fetchAndCheck() {
 
         @Override
         public synchronized void forward(KeyValue<T, ByteArray> target) {
-            if (state == State.COMPLETED) return;
-            else if (state == State.INIT) initialiseInternalIterator();
+            if (state == State.COMPLETED || ASC.isValidNext(target.key().bytes(), prefix.bytes())) return;
```

Comment:
Why is this of the form `isValidNext(to, from)` as opposed to `!isvalidNext(from, to)`?

Reply:
Checking if it is not valid you'd normally expect there to be an exception. So, mostly because I wanted to leave it "positive" to convey intent that this is OK behaviour.

Reply:
I see. What this tells me is that we need an `isValidPrev()` method. This version reads like "if it's valid (to do the opposite), then do nothing" which is more than a little confusing.
Alternatively, I'd argue that "if it isn't valid to move forward, don't move at all" reads quite a bit better, and it is a kind of error handling, as we can't legally forward to a value that comes before our current position.

Reply:
i see your point, I'll swap it around with the `!`